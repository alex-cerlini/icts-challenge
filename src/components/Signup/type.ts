export interface FormInputsProps {
  fullName: string;
  address: string;
  phone: string;
  email: string;
  birthDate: string;
}
