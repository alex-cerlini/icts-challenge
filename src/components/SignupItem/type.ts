import { UseFormRegister } from "react-hook-form";
import { SignupEnum } from "../Signup/enum";
import { FormInputsProps } from "../Signup/type";

export type SignupItemProps = {
  label: string;
  id: SignupEnum;
  register: UseFormRegister<FormInputsProps>;
};
