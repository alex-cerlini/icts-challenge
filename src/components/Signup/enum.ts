export enum SignupEnum {
  fullName = "fullName",
  address = "address",
  phone = "phone",
  email = "email",
  birthDate = "birthDate",
}
