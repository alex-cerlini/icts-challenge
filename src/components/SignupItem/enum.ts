export enum SignupItemTypeEnum {
  text = "text",
  date = "date",
  number = "number",
}
