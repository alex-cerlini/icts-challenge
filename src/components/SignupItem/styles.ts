import { styled, Box, BoxProps } from "@mui/material";

export const Container = styled(Box)<BoxProps>({
  marginBottom: 20,
  width: "88%",
  display: "flex",
  flexDirection: "column",
});
