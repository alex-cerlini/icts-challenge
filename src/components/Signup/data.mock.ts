import { SignupEnum } from "./enum";

export const fields = [
  {
    id: SignupEnum.fullName,
    label: "Nome completo",
  },
  {
    id: SignupEnum.address,
    label: "Endere√ßo",
  },
  {
    id: SignupEnum.phone,
    label: "Telefone",
  },
  {
    id: SignupEnum.email,
    label: "Email",
  },
  {
    id: SignupEnum.birthDate,
    label: "Data de nascimento",
  },
];
