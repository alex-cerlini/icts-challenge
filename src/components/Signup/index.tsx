import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import * as yup from "yup";
import { Box, Button } from "@mui/material";
import { FormInputsProps } from "./type";
import { fields } from "./data.mock";

import SignupItem from "../SignupItem";

import * as Style from "./styles";

const emailMask =
  /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;

const schema = yup
  .object({
    fullName: yup.string().required(),
    address: yup.string().required(),
    phone: yup.number().max(99999999999).required(),
    email: yup
      .string()
      .matches(emailMask, "E-mail inválido")
      .required("Campo obrigatório"),
    birthDate: yup.date().required("Data é obrigatório").min(new Date(0)),
  })
  .required();

const Signup = () => {
  const {
    register,
    handleSubmit,
    resetField,
    formState: { errors },
  } = useForm<FormInputsProps>({
    resolver: yupResolver(schema),
  });

  const onSubmit = (data: FormInputsProps) => {
    data.fullName = data.fullName.replace(/[0-9]/g, "");
    console.log(data);
  };

  const handleReset = () => {
    resetField("address");
    resetField("birthDate");
    resetField("email");
    resetField("fullName");
    resetField("phone");
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <Style.Container>
        {fields?.map((field) => (
          <SignupItem
            id={field.id}
            label={field.label}
            register={register}
            key={field.id}
          />
        ))}
        <Box
          style={{
            display: "flex",
            justifyContent: "space-between",
            width: "50%",
            marginTop: 18,
          }}
        >
          <Button
            variant="contained"
            type="submit"
            onClick={handleReset}
            color="error"
          >
            Limpar
          </Button>
          <Button variant="contained" type="submit">
            Enviar
          </Button>
        </Box>
      </Style.Container>
    </form>
  );
};

export default Signup;
