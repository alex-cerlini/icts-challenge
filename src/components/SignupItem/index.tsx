import { Input, InputLabel } from "@mui/material";
import { SignupItemProps } from "./type";
import * as Style from "./styles";
import { SignupEnum } from "../Signup/enum";
import { SignupItemTypeEnum } from "./enum";

const SignupItem = ({ label, id, register }: SignupItemProps) => {
  const typeInput = [
    SignupEnum.address,
    SignupEnum.email,
    SignupEnum.fullName,
  ].includes(id)
    ? SignupItemTypeEnum.text
    : id === SignupEnum.birthDate
    ? SignupItemTypeEnum.date
    : SignupItemTypeEnum.number;

  return (
    <Style.Container data-testid="container">
      <InputLabel htmlFor={id} role="label" data-testid="label">
        {label}
      </InputLabel>
      <Input id={id} {...register(id)} type={typeInput} />
    </Style.Container>
  );
};

export default SignupItem;
