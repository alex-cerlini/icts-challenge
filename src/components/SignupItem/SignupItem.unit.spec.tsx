import { render, screen } from "@testing-library/react";
import SignupItem from ".";
import { UseFormRegister } from "react-hook-form";
import { SignupEnum } from "../Signup/enum";
import { FormInputsProps } from "../Signup/type";

describe("SignupItem", () => {
  const register = jest.fn();
  const SignupString = "fullName";

  const renderSignupItem = () =>
    render(
      <SignupItem
        id={SignupEnum.fullName}
        label="Nome"
        register={register satisfies UseFormRegister<FormInputsProps>}
      />
    );
  it("Should render", () => {
    renderSignupItem();
    expect(screen.getByTestId("container")).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    const { baseElement } = renderSignupItem();
    expect(baseElement).toMatchSnapshot();
  });

  it("should render label on screen", () => {
    renderSignupItem();
    const label = screen.getByText("Nome");
    expect(label).toBeInTheDocument();
  });

  it("Should have input with correct id", () => {
    renderSignupItem();
    const label = screen.getByRole("textbox");
    expect(label.getAttribute("id")).toBe(SignupString);
  });

  it("should have label with correct for (equal input id)", () => {
    renderSignupItem();
    const label = screen.getByTestId("label");
    expect(label.getAttribute("for")).toBe(SignupString);
  });
});
