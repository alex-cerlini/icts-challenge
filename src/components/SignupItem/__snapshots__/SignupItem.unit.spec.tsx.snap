// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupItem should match snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-1rdoel6"
      data-testid="container"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
        data-testid="label"
        for="fullName"
        role="label"
      >
        Nome
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-q0jhri-MuiInputBase-root-MuiInput-root"
      >
        <input
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="fullName"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;
